cmd/common-methods.go

```diff
@@ -29,25 +29,32 @@ import (
 
 // Check if the target URL represents folder. It may or may not exist yet.
 func isTargetURLDir(targetURL string) bool {
-	targetURLParse := newClientURL(targetURL)
-	_, targetContent, err := url2Stat(targetURL)
-	if err != nil {
-		_, aliasedTargetURL, _ := mustExpandAlias(targetURL)
-		if aliasedTargetURL == targetURL {
-			return false
-		}
-		if targetURLParse.Path == string(targetURLParse.Separator) && targetURLParse.Scheme != "" {
-			return false
+	c, err := newClient(targetURL)
+	if err == nil {
+		if targetContent, serr := c.Stat(); serr == nil {
+			return targetContent.Type.IsDir()
 		}
-		if strings.HasSuffix(targetURLParse.Path, string(targetURLParse.Separator)) {
-			return true
+		// The target not found, we continue to guess according to bucket/prefix
+		// values in case of s3client
+		if s3c, ok := c.(*s3Client); ok {
```

Comment:
We shouldn't do these things, we are spilling over the details now. We have to validate this properly along with alias in picture Or make it part of the interface `IsDir()`  

Accessing internal functions like this leads to hard bugs.